Sources/Datadog/URLSessionAutoInstrumentation/URLSessionSwizzler.swift

```diff
@@ -0,0 +1,205 @@
+        )
+
+        private let method: FoundMethod
+        private let interceptor: URLSessionInterceptorType
+
+        init(interceptor: URLSessionInterceptorType) throws {
+            self.method = try Self.findMethod(with: Self.selector, in: URLSession.self)
+            self.interceptor = interceptor
+            super.init()
+        }
+
+        func swizzle() {
+            typealias Signature = @convention(block) (URLSession, URL, CompletionHandler?) -> URLSessionDataTask
+            swizzle(method) { previousImplementation -> Signature in
+                return { [weak self] session, url, completionHandler -> URLSessionDataTask in
+                    var taskReference: URLSessionDataTask?
+
+                    assert(completionHandler != nil)
+
+                    let newCompletionHandler: CompletionHandler = { data, response, error in
+                        if let task = taskReference { // sanity check, should always succeed
+                            self?.interceptor.taskCompleted(urlSession: session, task: task, error: error)
+                        }
+                        completionHandler?(data, response, error)
+                    }
+
+                    let task = previousImplementation(session, Self.selector, url, newCompletionHandler)
+                    taskReference = task
+
+                    self?.interceptor.taskCreated(urlSession: session, task: task)
+                    return task
+                }
+            }
+        }
+    }
+
+    /// Swizzles the `URLSession.dataTask(with:)` for `URLRequest`.
+    class DataTaskWithURLRequest: MethodSwizzler<
+        @convention(c) (URLSession, Selector, URLRequest) -> URLSessionDataTask,
+        @convention(block) (URLSession, URLRequest) -> URLSessionDataTask
+    > {
+        private static let selector = #selector(
+            URLSession.dataTask(with:) as (URLSession) -> (URLRequest) -> URLSessionDataTask
+        )
+
+        private let method: FoundMethod
+        private let interceptor: URLSessionInterceptorType
+
+        init(interceptor: URLSessionInterceptorType) throws {
+            self.method = try Self.findMethod(with: Self.selector, in: URLSession.self)
+            self.interceptor = interceptor
+            super.init()
+        }
+
+        func swizzle() {
+            typealias Signature = @convention(block) (URLSession, URLRequest) -> URLSessionDataTask
+            swizzle(method) { previousImplementation -> Signature in
+                return { [weak self] session, urlRequest -> URLSessionDataTask in
+                    let newRequest = self?.interceptor.modify(request: urlRequest) ?? urlRequest
+                    let task = previousImplementation(session, Self.selector, newRequest)
+                    if #available(iOS 13.0, *) {
+                        self?.interceptor.taskCreated(urlSession: session, task: task)
+                    }
```

Comment:
It is `URLSessionSwizzler's` responsibility to abstract the swizzlings and hide the complexity of `session.dataTask(...)` methods execution. The `interceptor` interface is simple:
```swift
protocol URLSessionInterceptor {
    func modify(request: URLRequest) -> URLRequest
    func taskCreated(..., task: URLSessionTask)
    func taskMetricsCollected(..., task: URLSessionTask, metrics:  URLSessionTaskMetrics)
    func taskCompleted(..., task: URLSessionTask, error: Error?)
}
```
and all the logic in `siwzzler` guarantees that each method of the `interceptor` is executed **only once** for each request send from the app. This makes adding logic to the `interceptor` very easy as we don't have to consider its methods being called for no reason. This is IMO the proper way of hidding complexity and implementation details in the right scope.

The `swizzler` tests assert that each method of `interceptor` is called only once per request (no matter of the iOS version and if called from Swift or Objective-C).

This also simplifies tests of the `interceptor` interactions with further components (Tracing handler + RUM Resource handler) as we don't need to test (e.g.) what happens to the `Span` or RUM Resource if the `interceptor.taskCreated()` is called twice.

If we consider duplicated calls in `interceptor`, we will result with way more possible control flows and unit tests.

Reply:
sorry maybe i wasn't clear enough
we should add `if interceptions[task] != nil` check only in `taskCreated` because it is where we add new interceptions. this adds 1 new test case, no?

Reply:
This will have following impact on existing tests:

```swift
// URLSessionSwizzlerTests:

func testWhenUsingTaskWithURL_itNotifiesCreationAndDoesNotModifiyTheRequest() {
   // If run prior to iOS 13.0, assert that the creation is notified twice.
}

// URLSessionInterceptorTests:

func testGivenTracerRegistered_whenInterceptingURLSessionTasks_itSendsSpanOnlyForFirstPartyRequests() {
   // Simulate calling completion several times, explain it with the comment 
   // why we assert that tracing handler is called only once.
}

func testGivenTracerNotRegistered_whenInterceptingURLSessionTasks_itDoesNotSendsSpanForAnyRequest() {
   // Simulate calling completion several times, explain it with the comment.
}
```
and require similar explanations/assertions for RUM Resources handler in next PR.

IMO _"this method can be called several times while the other are called only once"_ is a dangerous  design, not hiding the complexity in the right place and opening the component for other modifications which may weaken its contract (why not calling other methods several times as well in the future?). If we want to go this way, it should be somehow the `URLSessionSwizzler` checking which task creation was already notified and calling the `interceptor` only once. This is exactly what happens now, but with to explicit iOS version checks, instead of one-to-fit-all approach discussed in https://github.com/DataDog/dd-sdk-ios/pull/262#discussion_r500103775.

Reply:
okay 👌 i wanted to make sure this was considered too